src/geometry_aux.cpp

```diff
@@ -0,0 +1,569 @@
+  if (!file_exists(filename)) {
+    fatal_error("Geometry XML file '" + filename + "' does not exist!");
+  }
+
+  // Parse settings.xml file
+  pugi::xml_document doc;
+  auto result = doc.load_file(filename.c_str());
+  if (!result) {
+    fatal_error("Error processing geometry.xml file.");
+  }
+
+  // Get root element
+  pugi::xml_node root = doc.document_element();
+
+  // Read surfaces, cells, lattice
+  read_surfaces(root);
+  read_cells(root);
+  read_lattices(root);
+
+  // Allocate universes, universe cell arrays, and assign base universe
+  model::root_universe = find_root_universe();
+}
+
+//==============================================================================
+
+void
+adjust_indices()
+{
+  // Adjust material/fill idices.
+  for (auto& c : model::cells) {
+    if (c->fill_ != C_NONE) {
+      int32_t id = c->fill_;
+      auto search_univ = model::universe_map.find(id);
+      auto search_lat = model::lattice_map.find(id);
+      if (search_univ != model::universe_map.end()) {
+        c->type_ = FILL_UNIVERSE;
+        c->fill_ = search_univ->second;
+      } else if (search_lat != model::lattice_map.end()) {
+        c->type_ = FILL_LATTICE;
+        c->fill_ = search_lat->second;
+      } else {
+        std::stringstream err_msg;
+        err_msg << "Specified fill " << id << " on cell " << c->id_
+                << " is neither a universe nor a lattice.";
+        fatal_error(err_msg);
+      }
+    } else {
+      c->type_ = FILL_MATERIAL;
+      for (auto& mat_id : c->material_) {
+        if (mat_id != MATERIAL_VOID) {
+          auto search = model::material_map.find(mat_id);
+          if (search == model::material_map.end()) {
+            std::stringstream err_msg;
+            err_msg << "Could not find material " << mat_id
+                    << " specified on cell " << c->id_;
+            fatal_error(err_msg);
+          }
+          // Change from ID to index
+          mat_id = search->second;
+        }
+      }
+    }
+  }
+
+  // Change cell.universe values from IDs to indices.
+  for (auto& c : model::cells) {
+    auto search = model::universe_map.find(c->universe_);
+    if (search != model::universe_map.end()) {
+      c->universe_ = search->second;
+    } else {
+      std::stringstream err_msg;
+      err_msg << "Could not find universe " << c->universe_
+              << " specified on cell " << c->id_;
+      fatal_error(err_msg);
+    }
+  }
+
+  // Change all lattice universe values from IDs to indices.
+  for (auto& l : model::lattices) {
+    l->adjust_indices();
+  }
+}
+
+//==============================================================================
+//! Partition some universes with many z-planes for faster find_cell searches.
+
+void
+partition_universes()
+{
+  // Iterate over universes with more than 10 cells.  (Fewer than 10 is likely
+  // not worth partitioning.)
+  for (const auto& univ : model::universes) {
+    if (univ->cells_.size() > 10) {
```

Comment:
Can this 10 value change in the future ? does it need to be factored out as a variable ?